### PR TITLE
Support torch DDP distributed training

### PR DESCRIPTION
**Pull request by jacksonjacobs1**
This PR adds support for distributed training.

image

"cycle": a training repetition which completes after the train-pred loop hits the first `dist.barrier()`, resulting in the rotation of the pred_last and pred_latest tables.

Distributed training continues across 9 cycles without error:
image

And here we see that all patches receive predictions, showing that all workers complete their full shard subset before the cycle repeats:

image

- Workers process mutually exclusive subsets of shards.
- Each worker currently uses a single dataloader which naively iterates through every locally-available batch.
- Every batch is used for both prediction and backprop. Distributed training adds communication overhead per batch, as gradients are synced between workers. Distributed prediction may be substantially more efficient. Will need to look at torch profiling.
- Workers are not guaranteed to see the same number of batches per cycle. However, DDP requires that each worker syncs gradients the exact same number of times. There are two potential solutions to this:
  - `torch.distributed.Join` context manager
  -  pre-compute max_batches across all workers and modify the dataloader to yield empty batches until max_batches are reached. This is the solution I went with, happy to discuss.

## Next steps
Merge in latest DL model update and take a look at pytorch profiling.

### CODE REVIEW

**Comment by jacksonjacobs1** [comment]
max_batches is logged at the beginning of the train_worker loop for each worker, indicating that this test covers the edge case where workers see different numbers of batches:
```
[Worker 0] Shard counts: [1888, 2023, 1962, 1954, 1926, 1930, 1900, 2051, 1901, 1944, 1918, 1978, 1953, 1950, 1927, 1930], local batches: 33, max_batches (all-reduce): 33

[Worker 1] Shard counts: [1873, 1909, 1899, 1951, 1967, 1915, 1986, 1935, 1958, 1825, 1911, 1998, 1861, 1978, 1954, 1900], local batches: 32, max_batches (all-reduce): 33
```

File: docs/source/design_material_draft2/ray_technical_design.md [hunk 1]
```diff
@@ -1,155 +1,211 @@
...
+- `project_id` (int) — project to train for.
+- `app_config` (Dict[str, Any]) — read from application-level settings. Key values:
+  - `dl_patches_per_batch` — batch size for the dataloader.
+  - `patch_size` — image dimensions (H×W).
+  - `world_size` — total number of distributed workers, used to scale embedding coordinates.
+- `label_classes` (List[LabelClassResponse]) — valid (non-unassigned) label classes for the project.
+
+### LabelMap
+
+A `LabelMap` class provides bidirectional mapping between DB `label_class_id` and model class indices. The unassigned class (`label_class_id == 1`) is excluded from the model's output space entirely. Valid classes are sorted by `label_class_id` for deterministic mapping.
```
Comment (choosehappy) [inline docs/source/design_material_draft2/ray_technical_design.md:73]: this seems out of date?

File: docs/source/design_material_draft2/ray_technical_design.md [hunk 2]
```diff
@@ -1,155 +1,211 @@
...
+Pseudocode for the per-worker loop:
+
+```
+train_worker(config):
+    project_id = config["project_id"]
+    app_config = config["app_config"]
+    label_classes = config["label_classes"]
+    patches_per_batch = app_config.get("dl_patches_per_batch", 1000)
+    world_size = app_config.get("world_size", 4096)
+    GRID_SIZE_SCALE = world_size / GRID_SIZE
```
Comment (choosehappy) [inline docs/source/design_material_draft2/ray_technical_design.md:86]: this doesn't seem to be used here?

File: patchsorter/db/worker_client/patch.py [hunk 3]
```diff
@@ -115,6 +115,29 @@ def fetch_patches_by_shard(
...
+            List of ``COUNT(*)`` values, one per shard in *shard_ids*.
+        """
+        if not shard_ids:
+            return []
+        union_parts = " UNION ALL ".join(
+            f"SELECT {i} AS rn, COUNT(*) AS cnt FROM {build_table_name(self.project_id, sid)}"
+            for i, sid in enumerate(shard_ids)
+        )
+        result = self._session.execute(
+            text(f"SELECT cnt FROM ({union_parts}) sub ORDER BY rn"),
```
Comment (choosehappy) [inline patchsorter/db/worker_client/patch.py:137]: wow - thats...something isn't it?

File: patchsorter/dl/training.py [hunk 4]
```diff
@@ -337,20 +379,34 @@ def train_worker(config: Dict[str, Any]) -> None:
...
+        dataset = ShardDataset(worker_sm, project_id, assigned_shards, patches_per_batch, max_batches)
         for shard_id, batch in dataset:
+            # An empty batch is yielded when the shard subset is exhausted but max_batches has not been reached yet.
+            if not batch:
+                optimizer.zero_grad()
+                dummy_input = torch.zeros(2, 3, patch_size, patch_size, device=device)
+                with torch.autocast(device_type=device.type, dtype=torch.float16, enabled=True):
+                    z = backbone(dummy_input)
+                    emb, coords, logits = joint_head(z)
+                    dummy_loss = emb.sum() * 0.0 + coords.sum() * 0.0 + logits.sum() * 0.0
```
Comment (choosehappy) [inline patchsorter/dl/training.py:391]: is this striclty needed? like, why can't it just do "nothing" instead of doing "fake stuff"? i would point out that as written ,this would also have a consequence on the the inputs. e.g., batch norm doens't rely on gradients - its updated based on inputs, so since this is not in eval() mode, this forward pass will impact the BN nodes. perhaps minorly, but not great either way. then if one sets it to eval() this won't happen, but if this is in eval() mode, why bother to do the pass at all?

**Review by choosehappy — COMMENTED** [review]
there wasn't much in scope to review during this diff - mostly md files, but left a few comments in the code. will need to figure out a way to more intelligently review the bulk of the new code. i think you were going to create individual smaller PRs in specific places needing priortized review?

**Comment by choosehappy** [comment]
one thought - conceptually i'm wondering why there has to be a sync at all for the rotation? it seems like the rotation is actually quite shard specific? i.e., if worker 1 finishes its epoch consisteing of shard 1,2,3, while worker 2 finishes shard 4 and is working on shard 5 (s5...), why can't worker 1 internally rotate its table and start again at s1 while w2 continues to work on s5? since the batches are always local and mutually exclusive from other workers, there is actually no overarching requirement for a unified rotation, is there? that is to say, if someone queries, it is simply a union of the latest "current" table, which could be at a different "cycle" than the others? do  you see what i mean?

this would also remove the overall sync requirements and remove the need for the "empty batch" as well, i think?

**Comment by choosehappy** [comment]
> Distributed prediction may be substantially more efficient. 

yes , i expect so as well... if you think about it, it isn't really "distributed" in the same way the learning is - if the model is frozen, then it is pure local inference. this become embarassling parelllizable. training on the other hand requires backpop sync and update
